pages/DSNP/UserData.md

```diff
@@ -0,0 +1,84 @@
+An implementation MUST fail the Replace User Data Operation if the `etag` values supplied do not match the current `etag` values for all chunks of the specified data type.
+
+## Replace User Data Operation
+
+The Replace User Data Operation takes the following parameters:
+
+* The DSNP User ID of the user who controls the data
+  * Implementations MUST ensure that the principal invoking this Operation is this user or a transparent chain of delegation from the user to the principal exists.
+* A map containing the set of data types to update as the keys, and tuples consisting of (1) the schema version used to encode the data type, and (2) a list where each element includes a data chunk and its associated entity tag, as the values.
+
+If the Operation is successful, any previous data associated with the user for each data type included in the input MUST be removed and replaced by the new data.
+
+Data chunks should be generated for each included data type using the following sequence of operations:
+* Generate the data records, in Avro binary format, according to the versioned schema.
+* Divide the data records into one or more chunks, with each chunk containing a maximum byte count specified by the implementation.
+  * Note that the combined chunks constitute a different, and smaller, binary than a complete Avro file, as they exclude the header information, schema, and any block-level metadata.
+    Each chunk should start and end with a complete Avro object; in other words, chunk boundaries MUST NOT occur mid-record.
+    Applications SHOULD try to include as many records as possible within the byte length limit, but because compression and encryption may alter the byte size versus the raw Avro binary records, they may wish to use a heuristic approach that sometimes produces non-optimal chunks.
+* For each chunk generated, the application should then:
+  * If the data type requires compression, apply the compression codec noted.
+  * If the data type requires encryption,
+    * Retrieve the user's active (most recently announced) `keyAgreement` public key, U<sub>public</sub>.
+    If no key exists, one should be created and published as an Announcement before invoking the Operation.
+    * Create a sealed box, as in the [libsodium](https://doc.libsodium.org/public-key_cryptography/sealed_boxes) function `crypto_box_seal`, using U<sub>public</sub>.
+  * Include the previous `etag` value for the chunk. If the chunk is new, `etag` should be set to `null`.
+  If any chunks are to be deleted, they should be included in the input identified with the existing `etag` and a `null` value for the data.
+
+If the Operation is invoked successfully it MUST (synchronously) return a new set of `etag` values for each data type replaced, corresponding to the updated state of the data (with new chunks added and deleted chunks removed).
```

Comment:
`it` I'm not sure what the `it` here is. `it` will not be the implementation as Frequency or Ethereum are unable to respond with that data. It must be the DSNP library for that implementation?

That feels like a bit of an odd thing to require.

We could require that the async confirmation include that information, but I think that is all.

Reply:
I think of "the implementation" as inclusive of both the blockchain and the "adapter layer" (DSNP on Frequency). For example, Frequency will be defining in its compliance documentation the RPCs used to invoke DSNP Operations. Such an RPC could handle returning these values synchronously (which I think is the plan from discussion with @aramikm and @JoeCap08055). I can definitely change "it" to "the implementation" for clarity.